File: typescript/typelayer-http-server/bin/transport.js
"use strict";
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const typelayer_common_1 = require("typelayer-common");
const http = __importStar(require("http"));
function runHttp(settings, application) {
    const options = { port: settings.port };
    const handler = (nodeRequest, nodeResponse) => {
        return new Promise((resolve, reject) => {
            let body = [];
            nodeRequest
                .on("error", (e) => {
                reject(e);
            })
                .on("data", (chunk) => {
                body.push(chunk);
            })
                .on("end", () => {
                let binBody = Buffer.concat(body);
                let url = nodeRequest.url || "";
                let splitUrl = url.split("/");
                let request = new typelayer_common_1.HttpRequest({
                    path: splitUrl,
                    headers: typelayer_common_1.objectToMap(nodeRequest.headers),
                    body: binBody
                });
                let encoding = typelayer_common_1.getEncodingForRequest(request);
                application
                    .httpApplicationHandleRequest(request)
                    .then((appResponse) => {
                    nodeResponse.writeHead(appResponse.statusCode, typelayer_common_1.mapToObject(appResponse.headers));
                    appResponse.body.forEach(c => nodeResponse.write(c));
                    nodeResponse.end();
                    resolve();
                })
                    .catch(e => {
                    if (e instanceof typelayer_common_1.TypelayerException) {
                        let meta = typelayer_common_1.toSerializable(e.toMeta());
                        let encoder = encoding.encodingStartEncoding();
                        typelayer_common_1.dynamicPutExceptionMetaSerializable(encoder, meta);
                        nodeResponse.writeHead(400, {});
                        nodeResponse.end(encoder.encoderToBinary());
                    }
                    else {
                        reject(e);
                    }
                });
            });
        }).catch(e => {
            console.log(e);
            nodeResponse.writeHead(500, {});
            nodeResponse.end("Internal Error");
        });
    };
    http.createServer(handler).listen(options);
}
exports.runHttp = runHttp;

File: python/typelayer_http_server/typelayer_http_server/transport.py
from starlette.applications import Starlette
from starlette.requests import Request
from starlette.responses import Response
from typelayer_common.transports.http import (
    HTTPSettings,
    HTTPApplication,
    HTTPRequest,
    get_encoding_for_request,
    http_application_handle_request,
)


def make_asgi(settings: HttpSettings, application: HttpApplication):
    async def app(scope, receive, send):
        assert scope["type"] == "http"
        request = Request(scope, receive)
        split_url = request.url.path.split("/")
        loop = asyncio.get_event_loop()
        request_body = await request.body()
        request = HttpRequest(
            path=split_url, headers=request.headers, body=request_body
        )
        encoding = get_encoding_for_request(request)
        try:
            appResponse = loop.run_in_executor(
                None, app.http_application_handle_request, request
            )
        except typelayer_exception_common.TypelayerException as e:
            meta = to_serializable(e.to_meta())
            encoder = encoding.start_encoding()
            dynamic_put_exception_meta_serializable(encoder, meta)
            response = Response(
                content, status_code=400, media_type=encoding.mimetype()
            )
            await response(scope, receive, send)
            return
        except Exception as e:
            response = Response("Internal Error", status_code=500)
            await response(scope, receive, send)
            return

        content = b""
        for chunk in appResponse.body:
            content += chunk

        response = Response(
            content,
            status_code=200,
            media_type=encoding.mimetype(),
            headers=appResponse.headers,
        )
        await response(scope, receive, send)

    return app
